File: content/LLD/Reference-Code/1. OOPS Recap/1. Revision/abstraction.py
from abc import ABC, abstractmethod


class Shape(ABC):
    @abstractmethod
    def area(self):
        pass

    @abstractmethod
    def perimeter(self):
        pass


# Concrete Classes
class Rectangle(Shape):
    def __init__(self, length: int, breadth: int):
        self.length = length
        self.breadth = breadth

    def area(self):
        print(self.length * self.breadth)

    def perimeter(self):
        print(2 * (self.length + self.breadth))


r = Rectangle(5, 2)
r.area()
r.perimeter()

File: content/LLD/Reference-Code/1. OOPS Recap/1. Revision/Abstraction.java
abstract class Shape {
    public abstract void area();
    public abstract void perimeter();
}

// Concrete Classes
class Rectangle extends Shape {
    private int length;
    private int breadth;

    public Rectangle(int length, int breadth) {
        this.length = length;
        this.breadth = breadth;
    }

    @Override
    public void area() {
        System.out.println(this.length * this.breadth);
    }

    @Override
    public void perimeter() {
        System.out.println(2 * (this.length + this.breadth));
    }
}

public class Abstraction {
    public static void main(String[] args) {
        Rectangle r = new Rectangle(5, 2);
        r.area();
        r.perimeter();
    }
}
